rllib/env/multi_agent_episode.py

```diff
@@ -941,15 +937,15 @@ def cut(self, len_lookback_buffer: int = 0) -> "MultiAgentEpisode":
 
         # If there is hanging data (e.g. actions) in the agents' caches, we might have
         # to re-adjust the lookback len further into the past to make sure that these
-        # agents have at least one observation to look back to.
+        # agents have at least one observation to look back to. Otherwise, the timestep
+        # that got cut into will be "lost" for learning from it.
+        orig_len_lb = len_lookback_buffer
         for agent_id, agent_actions in self._hanging_actions_end.items():
             assert self.env_t_to_agent_t[agent_id].get(-1) == self.SKIP_ENV_TS_TAG
-            for i in range(1, self.env_t_to_agent_t[agent_id].lookback + 1):
-                if (
-                    self.env_t_to_agent_t[agent_id].get(-i, neg_index_as_lookback=True)
-                    != self.SKIP_ENV_TS_TAG
-                ):
-                    len_lookback_buffer = max(len_lookback_buffer, i)
+            for i in range(orig_len_lb, len(self.env_t_to_agent_t[agent_id].data) + 1):
```

Comment:
Why do we access here directly the raw `data` attribute when we have the `InfiniteLookbackBuffer` defined as a class with certain length methods?

Reply:
Great question and yes, this is a little hacky here. The len method returns the length but w/o(!) the lookback buffer.